doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,615 @@
+- Manager and host do not need to consider both paging and batching at
+  the same time, which is a complex convergence of concepts.
+
+#### Cons
+
+- Precludes batching optimisations by the manager when paging is
+  required.
+- Requires implementation of two methods by the manager: one for
+  batched, one for paged.
+- Potential difficulty for some hosts/queries to determine a priori
+  which API (paging or batch) is the most optimal choice.
+- Having the choice adds a foot-gun for host developers to shoot
+  themselves with by accident.
+
+### Batched paging option 2 - Whole-batch paging
+
+E.g. Page token
+
+```python
+# Callback taking index of batch element and selected page of results.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+    batch: List[EntityReference], relationship,
+    successCb: SuccessCallback, errorCb,
+    limit=-1, pageToken=None
+) -> AbstractPageToken: ...
+```
+
+E.g. Cursor object
+
+```python
+class AbstractPage:
+    def get(self) -> List[List[EntityReference]]:
+        ...
+    def next(self, pageSize):
+        ...
+
+def relatedEntities(
+    batch: List[EntityReference], relationship) -> AbstractPage: ...
+```
+
+#### Pros
+
+- Whole batch is paged at once, retaining batch-awareness in the
+  manager for subsequent pages.
+- Conceptually simpler API for both host and manager if only a single
+  page is active across the whole batch.
+
+#### Cons
+
+- Wasted processing of empty lists by the host when pages are exhausted
+  i.e. when there are still pages remaining for some inputs but not
+  others.
+- Batch elements cannot enter an error state in isolation, since paging
+  is whole-batch.
+
+### Batched paging option 3 - Per-element paging
+
+E.g. Page token
+
+```python
+# Callback taking index of batch element and selected page of results.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb,
+    limit: -1,
+    pageTokens: List[AbstractPageToken]
+): -> List[AbstractPageToken]: ...
+```
+
+E.g. Cursor object
+
+```python
+class AbstractPage:
+    def get(self) -> List[EntityReferences]:
+        ...
+    def next(self, pageSize):
+        ...
+
+# Callback taking index of batch element and cursor object.
+SuccessCallback = Callable[[int, AbstractPage], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb): ...
+```
+
+#### Pros
+
+- Manager is under no obligation to consider errors in other streams
+  when implementing query behaviour, as pagination method need not be synchronized
+
+#### Cons
+
+- Batching is very difficult to optimize for the manager, especially
+  after the first page.
+- Batch-optimized queries have unpredictable performance profile for
+  the host, given the wide range of potential implementations.
+
```

Comment:
[fix] The host cant express cursor modifications as a batch operation.

Reply:
I wonder if pros/cons that are specific to a particular paging strategy should go in a new section further down, and ideally only after we've rejected one or more of the options here, so we can reduce the matrix?

Reply:
Was just trying to keep it a bit more grokable, we know people get lost in these when there are a lot of sections.

Reply:
OK that's fine.  Though I'm now struggling to understand what your suggested Con means, please can you elaborate?

Reply:
Sorry was trying to describe how the cursor-per-element approach precludes the host from getting the next page for multiple cursors simultaneously (as they're distinct methods on each individual cursor) which breaks the batch. If you use tokens, then you can supply multiple tokens at once to the second invocation which keeps the batch-ness intact.

Reply:
Added a Con in 3619805a